File: src/advent_of_code/y2020/day15.py
data = list(map(int, open("resources/inputs/2020/day15.txt").read().split(",")))
spoken = {n: i for i, n in enumerate(data)}
n = data[-1]
prev = len(data) - 1
for i in range(prev + 1, 30_000_000):
    if i == 2020:
        print(n)
    n = i - 1 - prev
    prev = spoken.get(n, i)
    spoken[n] = i
print(n)

File: src/advent_of_code/y2020/day15.rs
// fn extend(vec: &mut Vec<i32>, n: usize) {
//     if vec.len() < n + 1 {
//         vec.resize(n + 1, -1);
//     }
// }

pub fn run(input: String) {
    let foo: Vec<usize> = input.split(",").map(|s| { s.trim_end().parse().unwrap() }).collect();
    let mut spoken = vec![0u32; 1 << 26];

    for (i, &n) in foo.iter().enumerate() {
        // extend(&mut spoken, n + 1);
        spoken[n] = (i + 1) as u32;
    }
    let mut n = *foo.last().unwrap();
    let mut prev = foo.len();

    for i in (prev + 1)..=30000000 {
        if i == 2020 + 1 { println!("{}", &n); }
        n = i - 1 - prev;
        // extend(&mut spoken, n + 1);
        prev = {
            let prev = spoken[n];
            if prev == 0 { i } else { prev as usize }
        };

        spoken[n] = i as u32;
    }
    println!("{}", &n);
}
